### DISCUSSION TITLE
Some functionality of --autotemplate seems broken?

### DISCUSSION THREAD

**Discussion by danpf** [twpayne/chezmoi#2147] [Q&A] [answered]
# Basic problem
from the docs it seems like doing chezmoi add --autotemplate should add .tmpl to the chezmoi file, but it doesn't seem to be doing that (--template is not implied?)

it also seems like  based on the issue below that --autotemplate should escape {{{ or {{ if it exists in the file, but it doesn't?

https://www.chezmoi.io/user-guide/manage-machine-to-machine-differences/
^ this says `chezmoi add --autotemplate` should  create dot_gitconfig.tmpl but i don't see that here with this test case

 https://github.com/twpayne/chezmoi/pull/723
^ this issue seems to say that autotemplate should add escapes for any present brackets, but that doesnt' seem to happen in my testcase

```chezmoi --version
chezmoi version v2.17.1, commit 565cbbe117746aa6bfec5f2cee20ae4cbbb5e645, built at 2022-05-30T10:24:34Z, built by goreleaser
```

# Background
I want to add some templating to my .zshrc file but it has a bunch of {{{ foldmarker blocks which causes chezmoi to die with the following error:
```chezmoi: template: dot_zshrc.tmpl:3: unexpected "{" in command```

Maybe I'm doing this wrong, or maybe it's not supported and I'm interpreting the docs incorrectly. either is fine I just wanted to check.

# tests
testcase:
```

rm -rf ~/.local/share/chezmoi/dot_simp*
echo "\"vim: set foldmethod=marker foldmarker={{,}}
{{ .machine_env }}
" > .simplefile
chezmoi add .simplefile --autotemplate --template
echo "printing chezmoi data*"
chezmoi execute-template '{{ .machine_env }}'
echo ""
echo "printing dot_simp*"
ls ~/.local/share/chezmoi/dot_simp*
cat ~/.local/share/chezmoi/dot_simp*


# THE ABOVE GIVES THIS:
printing chezmoi data*
y
printing dot_simp*
/home/me/.local/share/chezmoi/dot_simplefile.tmpl
"vim: set foldmethod=marker foldmarker={{,}}
{{ .machine_env }}


# REMOVING --template GIVES:
printing chezmoi data*
y
printing dot_simp*
/home/me/.local/share/chezmoi/dot_simplefile
"vim: set foldmethod=marker foldmarker={{,}}
{{ .machine_env }}
```

chezmoi config file:
```
[data]
    machine_env = "y"
```

**Comment by twpayne** [comment, marked as answer]
Thanks for reporting this. To be honest `--autotemplate` is pretty sketchy as the fundamental problem of "reversing" the output and template variables to get the input template has no clearly-defined friendly solution. For example, if you have a template variable (e.g. `machine_env`) is the single character `y` then `--autotemplate` should replace every `y` character in the entire file with `{{ .machine_env }}`, which is probably not what you want.

With regards to `--autotemplate` implying `--template`, `--autotemplate` only makes the added file a template if an actual variable substitution was found. This is for when you use `chezmoi add --autotemplate somedir/`, in which case it's desirable that only the files in `somedir/` that need to be templates end up as templates, rather than every single file in `somedir/` being a template. This is the case you're hitting in the reproduction case you give: no variable substitution is found, so the `.tmpl` suffix is not added.

The case of `--autotemplate` not automatically escaping `{{` and `}}` is definitely a deficiency. I'll fix that.

**Reply by danpf** [reply]
Oh I see, I was missinterpreting some parts of `--autotemplate` documentation.  What I was thinking was clearly wrong. 

I added some more documentation in a pr to clear that up (at least for me) https://github.com/twpayne/chezmoi/pull/2148/files

If that's not the place for it, or  if you don't like the changes you can just close it, nbd.
 
Thanks for responding so fast, I just started switching over to this and I think this project is great!
